### PR TITLE
Update version of JavaCPP to 1.4.2-SNAPSHOT

### LINKED ISSUE #5314: DevOps/Mono Repo: Move all version properties to parent pom
Now that we have a mono repo, we should align all dependency versions by moving the properties to the outer-most parent POM.
Currently, we have versions defined at the second-highest level (dl4j parent, nd4j parent, etc).

### PR DESCRIPTION
**Pull request by saudet**
Also consolidate version properties and profiles in parent pom.xml

Fixes #5314, #5324, and #5338

@sshepel @AlexDBlack Everything seems to be working OK for me, but please check it yourself before we merge this. So many things can break here...

### CODE REVIEW

**Comment by sshepel** [comment]
Add some changes, mostly formatting fixes and for small warnings.
Also remove wrong scm tags.
I would call all of these changes, phase 0. Phase 1, would be moving all dependencies (to parent pom.xml), updating explicitly provided plugin version with properties for the module root pom.xml files.
Phase 2, same as Phase 1 but for all children...

File: pom.xml [hunk 1]
```diff
@@ -63,17 +54,382 @@
...
+    <properties>
+        <maven.compiler.source>1.7</maven.compiler.source>
+        <maven.compiler.target>1.7</maven.compiler.target>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+
+        <deeplearning4j.version>${project.version}</deeplearning4j.version>
+        <dl4j.version>${project.version}</dl4j.version>
+        <nd4j.version>${project.version}</nd4j.version>
+        <datavec.version>${project.version}</datavec.version>
+        <dl4j-test-resources.version>${project.version}</dl4j-test-resources.version>
```
Comment (saudet) [inline pom.xml]: And we can't use project.version or project.parent.version here, that won't work for sure.

**Comment by AlexDBlack** [comment]
OK, so the build is fine, and after I pushed up a small lint fix.
However, it looks like test profiles need updating... I can't run datavec tests (```mvn clean test -Ptest-nd4j-native```). I think we just need to copy + paste this to a bunch of places:
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-core/pom.xml#L196-L203

Arbiter tests look OK. Haven't checked the rest.

File: deeplearning4j/pom.xml [hunk 2]
```diff
@@ -16,548 +18,29 @@
...
-                                    <goal>sign</goal>
-                                </goals>
-                            </execution>
-                        </executions>
-                    </plugin>
-                </plugins>
-            </build>
-        </profile>
-        <profile>
-            <id>release</id>
```
Comment (sshepel) [inline deeplearning4j/pom.xml]: This profile a bit differs from the one we provided in root pom.xml file.
@AlexDBlack how do you think is this crucial change?

**Comment by sshepel** [comment]
> I can't run datavec tests (mvn clean test -Ptest-nd4j-native)

I think problem here is related to enforcement rules in nd4j pom.xml https://github.com/deeplearning4j/deeplearning4j/blob/b5ec252abb387466f2e4f8823b44f35fa75caf67/nd4j/pom.xml#L288

**Comment by saudet** [comment]
@sshepel The tests for datavec were working, not perfectly, but anyway, something in your changes broke that...

**Comment by sshepel** [comment]
@saudet hm, I switched back to initial commit (`57fc78f`) and got the same error
```
[WARNING] Rule 0: org.apache.maven.plugins.enforcer.RequireActiveProfile failed with message:
Profile "testresources" is not activated.
```
@AlexDBlack is this the same error that you got?

**Comment by saudet** [comment]
@sshepel You added yourself the enforcer plugin to the parent pom.xml, so you'll need to add the workarounds required for that to work in the subprojects

**Comment by AlexDBlack** [comment]
@sshepel @saudet I've pushed up a fix for those datavec test profiles... tests now run (though I'm seeing a couple of failures - possibly unrelated to this PR?)

**Comment by saudet** [comment]
@AlexDBlack Tests from DataVec? I've just checked here and that's all green for me. Looks good to go!

**Comment by AlexDBlack** [comment]
After reviewing this again: I'm not yet happy to merge this, for two reasons:

1. First: we can't have the CUDA and CuDNN versions in the parent pom
2. ND4J is broken on Windows given the OpenBLAS update...

(Note also you'll need to rebase to master before you can build this branch)

Why can't we have the CUDA versions in the parent POM? Same reason we can't have scala versions in  the parent pom. That is: properties are not expanded when doing a build/install; they just refer to what is specified in the parent pom (which will be set to whatever it was during the last install).

Consider the following sequence of events:
1. Change CUDA version to 9.1
2. Build everything
3. Change CUDA version to 8.0
4. Build everything

After step 2: OK
deeplearning4j-cuda-9.1 uses nd4j-cuda-9.1 as a dependency (property is resolved from parent pom)

After step 4: Broken
Parent pom now has CUDA 8.0 as the property
deeplearning4j-cuda-9.1 uses nd4j-cuda-8.0 as a dependency (property is resolved from parent pom)
deeplearning4j-cuda-8.0 uses nd4j-cuda-8.0 as a dependency (property is resolved from parent pom)

---

As for the Windows error:

```
java.lang.UnsatisfiedLinkError: no jniopenblas in java.library.path
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:1221)
	at org.bytedeco.javacpp.Loader.load(Loader.java:980)
	at org.bytedeco.javacpp.Loader.load(Loader.java:879)
	at org.bytedeco.javacpp.openblas.<clinit>(openblas.java:10)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.bytedeco.javacpp.Loader.load(Loader.java:938)
	at org.bytedeco.javacpp.Loader.load(Loader.java:895)
	at org.bytedeco.javacpp.presets.openblas.blas_set_num_threads(openblas.java:174)
	at org.nd4j.linalg.cpu.nativecpu.blas.CpuBlas.setMaxThreads(CpuBlas.java:120)
	at org.nd4j.nativeblas.Nd4jBlas.<init>(Nd4jBlas.java:35)
	at org.nd4j.linalg.cpu.nativecpu.blas.CpuBlas.<init>(CpuBlas.java:15)
	at org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory.createBlas(CpuNDArrayFactory.java:94)
	at org.nd4j.linalg.factory.BaseNDArrayFactory.blas(BaseNDArrayFactory.java:68)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.getEnvironmentInformation(NativeOpExecutioner.java:1257)
	at org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner.printEnvironmentInformation(DefaultOpExecutioner.java:652)
	at org.nd4j.linalg.factory.Nd4j.initWithBackend(Nd4j.java:6932)
	at org.nd4j.linalg.factory.Nd4j.initContext(Nd4j.java:6802)
	at org.nd4j.linalg.factory.Nd4j.<clinit>(Nd4j.java:213)
	at org.deeplearning4j.gradientcheck.CNNGradientCheckTest.<clinit>(CNNGradientCheckTest.java:45)
Caused by: java.lang.UnsatisfiedLinkError: C:\Users\Alex\.javacpp\cache\openblas-0.3.0-1.4.2-20180602.224453-12-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\jniopenblas.dll: The specified procedure could not be found
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
	at java.lang.Runtime.load0(Runtime.java:809)
	at java.lang.System.load(System.java:1086)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:1201)
```

**Comment by saudet** [comment]
Ah yes, I knew there was a reason to have the CUDA versions where they were, oops. We'll add a comment for that so we don't forget again.

OpenBLAS is working for me on Windows with MKL, I tested that case specifically. Which version of MKL are you using, just to make sure we're using the same?

**Comment by AlexDBlack** [comment]
MKL version 2018.1.156

```
c:\>where mkl_rt.dll
C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2018.1.156\windows\redist\intel64_win\mkl\mkl_rt.dll
```

**Comment by saudet** [comment]
That's the same version I've tested here, and I've tested it again and it's still working fine. Could you remove `~/.javacpp/cache` and give me the output with `-Dorg.bytedeco.javacpp.logger.debug`?

**Comment by saudet** [comment]
Should be fixed with commit https://github.com/bytedeco/javacpp-presets/commit/daf57f71be0cd988fe51d96fbd09b36b93eaa08f, try again with `mvn -U ...`. It's probably loading MKL 2017 on your machine. It tries to load from `C:/Program Files (x86)/IntelSWTools/compilers_and_libraries/windows/redist/` before looking in the PATH, to prevent loading random DLLs from Anaconda, etc.

**Comment by AlexDBlack** [comment]
@saudet I tested on another windows machine with 2016.3.207 and it's fine there... so could just be old cache.
Will double check on my main machine tomorrow (switching branches + rebuilding is kind of disruptive to what I'm working on at present :))

**Comment by AlexDBlack** [comment]
@saudet OK, so I've tested again on the PC I had issues with, after clearing the cache: no more crash.
However, it's not loading MKL (according to ND4J init logging) under either maven or intellij.
The exact same PC (no messing with paths etc) does load MKL fine under 1.0.0-beta.

Here's the output: (using -Dorg.bytedeco.javacpp.logger.debug -Dorg.bytedeco.javacpp.logger=slf4j)
https://gist.github.com/AlexDBlack/624ad520d6a6222640ec3b8178f6f980

As for the MKL loading issue, maybe we merge this PR and deal with that separately...

**Comment by saudet** [comment]
You're getting a lot of weird messages in the log there... I'll try to reproduce that here, but will let you know if I can't. Yes, let's merge this PR if it's working well enough for now.

**Comment by AlexDBlack** [comment]
@saudet yeah, everything else seems OK - I'm happy to merge this.

**Comment by saudet** [comment]
@AlexDBlack It looks like you've enabled symbolic links on your account. It wasn't working with that on Windows, so I've fixed this in commit https://github.com/bytedeco/javacpp-presets/commit/612b3972e10d9f3a0cec9485e63d57720d6ca2f6, but it wasn't working before either... Anyway, let me know if it's still not picking up MKL with this (snapshots are up as of a couple of minutes ago). You might need to trash `~/.javacpp/cache` to get it working.

**Pull request merged**